runtime/src/accounts_db.rs

```diff
@@ -1266,6 +1266,43 @@ impl AccountsDB {
         }
     }
 
+    pub fn remove_unrooted_slot(&self, remove_slot: Slot) {
+        if self.accounts_index.read().unwrap().is_root(remove_slot) {
+            panic!("Trying to remove accounts for rooted slot {}", remove_slot);
+        }
+
+        // Remove old bank hash
+        self.bank_hashes.write().unwrap().remove(&remove_slot);
+
+        let pubkey_sets: Vec<HashSet<Pubkey>> = self.scan_account_storage(
+            remove_slot,
+            |stored_account: &StoredAccount, _, accum: &mut HashSet<Pubkey>| {
+                accum.insert(stored_account.meta.pubkey);
+            },
+        );
+
+        // Purge this slot from the accounts index
+        {
+            let accounts_index = self.accounts_index.read().unwrap();
+            for pubkey_set in pubkey_sets {
+                for pubkey in pubkey_set {
+                    accounts_index
+                        .account_maps
+                        .get(&pubkey)
+                        .expect("account in appendvec must have entry in accounts index")
+                        .1
```

Comment:
@sakridge this is where I'm removing from the accounts index. The type of each value in the map  is `AccountMapEntry<T> = (AtomicU64, RwLock<SlotList<T>>);`. 

 From what I gathered, the first item `.0` in the tuple is actually a refcount of sorts, that gets decremented in  `unref_from_storage` which looks at the `AppendVec` data, so it seems like I don't have to touch it here in the purge logic, let me know if I actually do need to handle it :D

Reply:
Uh, yea I would expect you have to handle it.

Reply:
It's a count of how many times that account appears in the storage. I guess if you delete the append_vec, on the cleanup it goes and decrements the counts of all accounts contained in that append_vec.

Reply:
Seems this is the same code as `AccountsDB::process_dead_slots`, can we just call that path?

Reply:
@sakridge, good point, I think there's a lot of overlap. `process_dead_slots` covers:

1) Clearing bank hashes here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1780https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1780

2) Purging the storage index here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1265

However, I don't think it goes through the accounts index and purges the slot entries for each pubkey in that index like I'm doing here: https://github.com/solana-labs/solana/pull/9805/files#diff-2099c5256db4eb5975c8834af38f6456R1287-R1296, it seems to only call https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1774, which only removes the slot from the list of roots. I'm assuming this works because it's assumed by this point no further children will build on top of/reference the accounts of this "dead" slot.

Thus it seems like I can call the `process_dead_slots` to purge the bank hash and the storage index, and then additionally filter the accounts index separately like I'm doing here: https://github.com/solana-labs/solana/pull/9805/files#diff-2099c5256db4eb5975c8834af38f6456R1287-R1296. Does that sound like an accurate assessment?

Reply:
Yea, that sounds right.

Reply:
Updated!